src/lib/comms/sol-lwm2m-security.c

```diff
@@ -0,0 +1,470 @@
+/*
+ * This file is part of the Soletta (TM) Project
+ *
+ * Copyright (C) 2016 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#define SOL_LOG_DOMAIN &_lwm2m_security_domain
+
+#include "sol-log-internal.h"
+#include "sol-util-internal.h"
+#include "sol-coap.h"
+#include "sol-lwm2m-security.h"
+#include "sol-platform.h"
+#include "sol-socket-dtls.h"
+#include "sol-socket.h"
+#include "sol-str-slice.h"
+#include "sol-util-internal.h"
+#include "sol-vector.h"
+
+SOL_LOG_INTERNAL_DECLARE(_lwm2m_security_domain, "lwm2m-security");
+
+#ifdef DTLS
+
+struct sol_socket *sol_coap_server_get_socket(const struct sol_coap_server *server);
+
+static ssize_t
+get_psk_from_server_or_bs_server(const void *data, struct sol_str_slice id,
+    char *psk, size_t psk_len)
+{
+    const struct sol_lwm2m_security *ctx = data;
+    struct sol_lwm2m_server *lwm2m_server;
+    struct sol_lwm2m_bootstrap_server *lwm2m_bs_server;
+    struct sol_vector *known_psks;
+    struct sol_lwm2m_security_psk *stored_psk;
+    uint16_t i;
+
+    SOL_DBG("Looking for PSK with ID=%.*s", SOL_STR_SLICE_PRINT(id));
+
+    //If the caller expects a PSK_LEN less than 16 bytes
+    if (psk_len < SOL_DTLS_PSK_KEY_LEN)
+        return -ENOBUFS;
+
+    if (ctx->type == LWM2M_SERVER) {
+        SOL_DBG("ctx->type = LWM2M_SERVER");
+        lwm2m_server = ctx->entity;
+        known_psks = lwm2m_server->known_psks;
+    } else {
+        SOL_DBG("ctx->type = LWM2M_BOOTSTRAP_SERVER");
+        lwm2m_bs_server = ctx->entity;
+        known_psks = lwm2m_bs_server->known_psks;
+    }
+
+    SOL_VECTOR_FOREACH_IDX (known_psks, stored_psk, i) {
+        if (sol_str_slice_eq(sol_str_slice_from_blob(stored_psk->id), id)) {
+            if (stored_psk->key->size != SOL_DTLS_PSK_KEY_LEN) {
+                SOL_WRN("The PSK '%.*s' is %zu-bytes long; expecting a %d-bytes long PSK",
+                    SOL_STR_SLICE_PRINT(sol_str_slice_from_blob(stored_psk->key)),
+                    stored_psk->key->size, SOL_DTLS_PSK_KEY_LEN);
+                return -EINVAL;
+            }
+
+            memcpy(psk, stored_psk->key->mem, stored_psk->key->size);
+            return (ssize_t)SOL_DTLS_PSK_KEY_LEN;
```

Comment:
`stored_psk->key->size` MUST be equal to `SOL_DTLS_PSK_KEY_LEN` (see answer above an line 67), since the PSK Key has a fixed size. So, basically, there's no difference... I just think it's a little better to read it this way.